packages/back-end/src/controllers/stripe.ts

```diff
@@ -1,144 +1,168 @@
-const stripe = new Stripe(STRIPE_SECRET || "", { apiVersion: "2020-08-27" });
-
-async function updateSubscription(subscription: string | Stripe.Subscription) {
-  // Make sure we have the full subscription object
-  if (typeof subscription === "string") {
-    subscription = await stripe.subscriptions.retrieve(subscription);
-  }
-
-  const stripeCustomerId =
-    typeof subscription.customer === "string"
-      ? subscription.customer
-      : subscription.customer.id;
-
-  await updateOrganizationByStripeId(stripeCustomerId, {
-    subscription: {
-      id: subscription.id,
-      qty: subscription.items.data[0].quantity || 1,
-      trialEnd: subscription.trial_end
-        ? new Date(subscription.trial_end * 1000)
-        : null,
-      status: subscription.status,
-    },
-  });
-}
-export async function postStartTrial(
-  req: AuthRequest<{ qty: number; name: string }>,
+  getNumberOfMembersAndInvites,
+  getOrgFromReq,
+} from "../services/organizations";
+import {
+  updateSubscriptionInDb,
+  stripe,
+  getCoupon,
+  getPrice,
+  getStripeCustomerId,
+} from "../services/stripe";
+import { SubscriptionQuote } from "../../types/organization";
+
+export async function postNewSubscription(
+  req: AuthRequest<{ qty: number }>,
   res: Response
 ) {
-  const { qty, name } = req.body;
+  const { qty } = req.body;
 
-  // If user already has a subscription, return immediately
-  if (req.organization?.subscription?.id) {
-    return res.status(200).json({
-      status: 200,
-    });
+  req.checkPermissions("organizationSettings");
+
+  const { org } = getOrgFromReq(req);
+
+  const desiredQty = getNumberOfMembersAndInvites(org);
+
+  if (desiredQty !== qty) {
+    throw new Error(
+      "Number of users is out of date. Please refresh the page and try again."
+    );
   }
 
-  try {
-    // Create organization first if needed
-    if (!req.organization) {
-      if (name.length < 3) {
-        throw new Error("Company name must be at least 3 characters long");
-      }
-      req.organization = await createOrganization(
-        req.email || "",
-        req.userId || "",
-        name,
-        ""
-      );
-    }
+  const stripeCustomerId = await getStripeCustomerId(org);
 
-    // Create customer in Stripe if not exists
-    if (!req.organization.stripeCustomerId) {
-      const resp = await stripe.customers.create({
-        email: req.email || "",
-        name: req.name || "",
-        metadata: {
-          user: req.userId || "",
-          organization: req.organization.id,
-        },
-      });
-      req.organization.stripeCustomerId = resp.id;
-
-      await updateOrganization(req.organization.id, {
-        stripeCustomerId: resp.id,
-      });
+  const existingSubscriptions = await stripe.subscriptions.list({
+    customer: stripeCustomerId,
+  });
+
+  existingSubscriptions.data.forEach((subscription) => {
+    if (subscription.status === ("active" || "past_due" || "trialing")) {
+      updateSubscriptionInDb(subscription);
```

Comment:
@jdorn Is this the best way to do this? I originally was awaiting this, and kept getting some unhandled promise rejections. In my testing, it works as expected, but curious if there is a better alternative.

Reply:
Async/await and Promises don't work well with `forEach` loops.  `forEach` runs a function for every array element immediately without waiting for async operations to finish. `map` is similar, but it gives you back return values, which you can use to await everything at the end.  You could also use a normal `for` loop, which lets you just use await directly without anything special.

```ts
// Runs one iteration at a time, waiting for async operations to finish before continuing the loop
for (let i=0; i<items.length; i++) {
  await somethingAsync(items[i]);
}

// Runs all iterations in parallel immediately and waits at the end for everything to finish
const promises = items.map(async (item) => {
  await somethingAsync(item);
});
await Promise.all(promises);

// map approach, but as a one-liner without an extra 'promises' variable
await Promise.all(items.map(async (item) => {
  await somethingAsync(item);
}));
```

The `for` approach is simpler and behaves how you would expect it to, but can be slow since it only runs one iteration at a time.

The `map` approach is a little confusing if you aren't used to the pattern, but can be much faster since it runs things in parallel.  You do have to be careful though if your array is large. You don't want to run 1000 Mongo queries all at once for example.

In this case, the number of subscriptions for someone is always going to be small, so `map` is fine.  I try to avoid `for` loops when possible.